### PR TITLE
feat: Port code to dart

### PR DESCRIPTION
**Pull request by Lee31416**
GitHub Issue: #

## Proposed Changes

 - [ ] Bug fix
 - [x] Feature
 - [ ] Code style update (formatting)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build or CI related changes
 - [ ] Documentation content changes
 - [ ] Other, please describe:

I'm porting the review service to dart in this pr.

## What is the current behavior?

N/A

## What is the new behavior?

N/A

## Checklist

Please check that your PR fulfills the following requirements:

- [x] Documentation has been added/updated.
- [x] Automated tests for the changes have been added/updated.
- [x] Updated [CHANGELOG.md](../CHANGELOG.md).

### CODE REVIEW

File: test/src/review_service_test.mocks.dart
Comment (takla21) [inline test/src/review_service_test.mocks.dart:1]: Should that file be in the repo?

File: test/src/memory_review_settings_source_test.dart [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import 'package:flutter_test/flutter_test.dart';
+import 'package:review_service/src/review_service/review_settings.dart';
+import 'package:review_service/src/review_service/memory_review_settings_source.dart';
+
+void main() {
+  test('MemoryReviewSettingSource should not return null when reading from a new instance', () async {
+    // Arrange
```
Comment (takla21) [inline test/src/memory_review_settings_source_test.dart:15]: the tabs here are not consistent

File: test/src/review_service_test.mocks.dart
Comment (Lee31416) [reply test/src/review_service_test.mocks.dart:1]: no

File: test/src/memory_review_settings_source_test.dart [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    var source = MemoryReviewSettingsSource<ReviewSettings>();
+    var value = ReviewSettings(requestCount: 10);
+
+    await source.write(value);
+
+    // Act
+    var result = await source.read();
+
+    // Assert
+    expect(value, result);
```
Comment (takla21) [inline test/src/memory_review_settings_source_test.dart]: I think you want the first parameter to be the actual result and the second parameter to be the expected result, so if it fails, the error message makes sense

File: lib/src/review_service/review_service.extensions.dart [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+              firstApplicationLaunch: DateTime.now(),
+              applicationLaunchCount: reviewSettings.applicationLaunchCount + 1,
+            ) as TReviewSettings
+          : reviewSettings.copyWith(
+              applicationLaunchCount: reviewSettings.applicationLaunchCount + 1,
+            ) as TReviewSettings;
+    });
+  }
+
+  Future<void> trackPrimaryActionCompleted() async {
```
Comment (takla21) [inline lib/src/review_service/review_service.extensions.dart:22]: I think it misses documentation for those extension method

File: lib/src/review_service/review_conditions_builder.extensions.dart [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    ));
+    return this;
+  }
+
+  IReviewConditionsBuilder<TReviewSettings> custom(bool Function(TReviewSettings, DateTime) condition) {
+    conditions.add(SynchronousReviewCondition<TReviewSettings>(condition));
+    return this;
+  }
+
+  IReviewConditionsBuilder<TReviewSettings> customAsync(
```
Comment (takla21) [inline lib/src/review_service/review_conditions_builder.extensions.dart]: Missing documentation on those methods as well

File: README.md [hunk 5]
```diff
@@ -1,39 +1,317 @@
-# Open Source Project Template
+# Review Service
 
-This repository contains a template to seed a repository for an Open Source
-project.
+This repository introduces abstractions around native review capabilities to ease code sharing and testability.
+It also introduces business logic to quickly configure conditions and state tracking to prompt for reviews at the right moment.
 
-## How to use this template
+[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](LICENSE) ![Version](https://img.shields.io/nuget/v/ReviewService.Abstractions?style=flat-square) ![Downloads](https://img.shields.io/nuget/dt/ReviewService.Abstractions?style=flat-square)
```
Comment (takla21) [inline README.md]: I don't think you want to reference the nuget (.NET version) ones here

Comment (Lee31416) [reply README.md]: Right I forgot the todos

File: CHANGELOG.md [hunk 6]
```diff
@@ -10,3 +10,4 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
 ## 0.3.0
 
 - Migrated the pipeline from the .NET version of the package.
+- Ported the code from the .Net version of the package to dart.
```
Comment (Soap-141) [inline CHANGELOG.md]: `.NET`.

File: lib/src/review_service/memory_review_settings_source.dart [hunk 7]
```diff
@@ -0,0 +1,19 @@
+import 'package:review_service/src/review_service/review_settings.dart';
+import 'package:review_service/src/review_service/review_settings_source_interface.dart';
+
+/// In-memory implementation of [IReviewSettingsSource].
+final class MemoryReviewSettingsSource<TReviewSettings extends ReviewSettings> implements IReviewSettingsSource<TReviewSettings> {
+  TReviewSettings _reviewSettings = const ReviewSettings() as TReviewSettings; // TODO: See if this is good. In C# it's default.
```
Comment (Soap-141) [inline lib/src/review_service/memory_review_settings_source.dart]: Don't forget this `TODO`.

File: lib/src/review_service/review_conditions_builder.extensions.dart
Comment (Soap-141) [inline lib/src/review_service/review_conditions_builder.extensions.dart:1]: No documentation on all those extensions?

File: lib/src/review_service/review_prompter_interface.dart [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+  ReviewPrompter({
+    required Logger logger,
+  })  : _logger = logger;
+
+  @override
+  Future<void> tryPrompt() async {
+    _logger.d('Trying to prompt user to review the current application.');
+
+    final InAppReview inAppReview = InAppReview.instance;
```
Comment (Soap-141) [inline lib/src/review_service/review_prompter_interface.dart]: This could probably be a private field or you could just do InAppReview.instance.requestReview(); since it is the only time you use it

File: lib/src/review_service/review_service.dart [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    required Logger logger,
+    required IReviewPrompter reviewPrompter,
+    required IReviewSettingsSource<TReviewSettings> reviewSettingsSource,
+    required IReviewConditionsBuilder<TReviewSettings> reviewConditionsBuilder,
+  })  : _logger = logger,
+        _reviewPrompter = reviewPrompter,
+        _reviewSettingsSource = reviewSettingsSource,
+        _reviewConditions = reviewConditionsBuilder.conditions;
+
+  Future<void> _trackReviewRequested() async {
```
Comment (Soap-141) [inline lib/src/review_service/review_service.dart:49]: Does this have documentation in the C# package?

File: lib/src/review_service/review_prompter_interface.dart [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+    if (await inAppReview.isAvailable()) {
+        inAppReview.requestReview();
+    }
+
+    _logger.i('Prompted user to review the current application.');
```
Comment (jeanplevesque) [inline lib/src/review_service/review_prompter_interface.dart]: Shouldn't this be inside the `if`?
Also, shouldn't we have a different log message when `isAvailable()` returns false?

File: lib/src/review_service/review_settings.dart
Comment (Soap-141) [inline lib/src/review_service/review_settings.dart:1]: I would remove the 'Gets or sets' in each comment, those are not properties. They are 'final' fields. I would use the part after.

File: pubspec.yaml [hunk 11]
```diff
@@ -10,8 +10,14 @@ environment:
...
+  in_app_review: ^2.0.9
+  logger: ^2.3.0
 
 dev_dependencies:
   flutter_test:
     sdk: flutter
   lints: ^3.0.0
+  mockito: ^5.4.4
+  build_runner: ^2.4.8
+  get_it: ^7.7.0
```
Comment (Soap-141) [inline pubspec.yaml]: Where is `get_it` used? I'm sorry, may be I missed it?

File: test/src/review_service_test.dart [hunk 12]
```diff
@@ -1,10 +1,79 @@
...
+import 'package:review_service/src/review_service/review_service.dart';
 
+import 'review_service_test.mocks.dart';
+
+@GenerateNiceMocks([MockSpec<IReviewPrompter>()])
 void main() {
-  group('ReviewService', () {
-    test('can be instantiated', () {
-      expect(ReviewService(), isNotNull);
+  const int maxInteger =  0x7FFFFFFFFFFFFFFF;
```
Comment (Soap-141) [inline test/src/review_service_test.dart:16]: Small comment here to explain what this is.

File: test/src/review_service_test.mocks.dart
Comment (Soap-141) [reply test/src/review_service_test.mocks.dart:1]: Update git ignore, see Flutter App Template

**Review by Soap-141 — COMMENTED** [review]
Don't forget to properly fill the PR template, you can add N/A to the current behavior or remove it.

File: lib/src/review_service/review_condition_interface.dart [hunk 13]
```diff
@@ -0,0 +1,10 @@
+import 'package:review_service/src/review_service/review_settings.dart';
+
+/// A condition used to determine if a review should be requested based on [ReviewSettings].
+abstract interface class IReviewCondition<TReviewSettings extends ReviewSettings> {
```
Comment (jeanplevesque) [inline lib/src/review_service/review_condition_interface.dart]: Shouldn't we remove the capital I at the start of interface names? It seems like it isn't the standard in Dart?

Comment (Soap-141) [reply lib/src/review_service/review_condition_interface.dart]: We can, like we did on the Flutter Application Template.

File: pubspec.yaml (hunk 11 above)
Comment (Lee31416) [reply pubspec.yaml]: I added it to test something for the readme and forgot to remove it

File: build/test_cache/build/cache.dill.track.dill
Comment (takla21) [inline build/test_cache/build/cache.dill.track.dill:1]: what are the files under `test_case`, `unit_test_assets` & `shaders`?

File: lib/src/review_service/logging_review_prompter.dart [hunk 14]
```diff
@@ -0,0 +1,17 @@
+import 'package:logger/logger.dart';
+import 'package:review_service/src/review_service/review_prompter.dart';
+
+/// Implementation of [ReviewPrompter] that logs information.
+final class LoggingReviewPrompter implements ReviewPrompter {
+  final Logger logger;
+
+  /// <summary>
```
Comment (takla21) [inline lib/src/review_service/logging_review_prompter.dart]: Hum I'm not sure it's a valid format for dart documentation

File: build/test_cache/build/cache.dill.track.dill
Comment (Lee31416) [reply build/test_cache/build/cache.dill.track.dill:1]: I dont know i deleted them but they somehow came back

File: lib/src/review_service/logging_review_prompter.dart (hunk 14 above)
Comment (takla21) [reply lib/src/review_service/logging_review_prompter.dart]: @Soap-141

Comment (Lee31416) [reply lib/src/review_service/logging_review_prompter.dart]: shit I forgot to remove the tags

File: lib/src/review_service/review_condition_builder.dart [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+
+  static ReviewConditionsBuilder<TReviewSettings>
+      defaultBuilder<TReviewSettings extends ReviewSettings>() {
+    return ReviewConditionsBuilderImplementation<TReviewSettings>()
+      ..minimumApplicationLaunchCount(3)
+      ..minimumElapsedTimeSinceApplicationFirstLaunch(const Duration(days: 5))
+      ..minimumPrimaryActionsCompleted(2)
+      ..minimumElapsedTimeSinceLastReviewRequest(const Duration(days: 15));
+  }
+
```
Comment (takla21) [inline lib/src/review_service/review_condition_builder.dart]: useless empty-line here

File: build/test_cache/build/cache.dill.track.dill
Comment (takla21) [reply build/test_cache/build/cache.dill.track.dill:1]: They should probably be part of the gitignore

**Review by takla21 — APPROVED** [review]

File: lib/src/review_service/logging_review_prompter.dart (hunk 14 above)
Comment (Soap-141) [reply lib/src/review_service/logging_review_prompter.dart]: @takla21 is right

**Review by Soap-141 — APPROVED** [review]

File: README.md [hunk 16]
```diff
@@ -1,39 +1,318 @@
...
-## Features (to keep as-is, configure or remove)
-- [Mergify](https://mergify.io/) is configured. You can edit or remove [.mergify.yml](/.mergify.yml).
+Before getting started, please read the [Android](https://developer.android.com/guide/playcore/in-app-review) and [iOS](https://developer.apple.com/design/human-interface-guidelines/ratings-and-reviews) application review documentation.
 
-The following is the template for the final README.md file:
+## Getting Started
 
----
+1. Add the `ReviewService` package to your project.
+   > 💡 If you need to implement more platforms or create custom implementations, you can use the `ReviewService.Abstractions` package.
```
Comment (jeanplevesque) [inline README.md]: I don't think `ReviewService.Abstractions` is applicable to the dart packages.

File: README.md [hunk 17]
```diff
@@ -1,39 +1,318 @@
...
+Here is a simple code that does dependency injection using `get_it`.
+
+```dart
+   GetIt getIt = GetIt.instance;
+
+   // Register the ReviewPrompter implementation with GetIt
+   getIt.registerSingleton<IReviewPrompter>(ReviewPrompter());
+```
+
+> 💡 We recommend that you define your own interface that wraps `IReviewService<YouChoiceOfReviewSettings>` to make the usage code leaner and ease any potential refactorings.
```
Comment (jeanplevesque) [inline README.md]: The interface names don't match the code.

**Review by jeanplevesque — CHANGES_REQUESTED** [review]

File: lib/src/review_service/review_service.extensions.dart [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+import 'package:review_service/src/review_service/review_settings.dart';
+
+/// Extensions of [IReviewService].
+extension ReviewServiceExtensions<TReviewSettings extends ReviewSettings> on ReviewService<TReviewSettings> {
+
+  /// Tracks that the application was launched.
+  Future<void> trackApplicationLaunched() async {
+    await updateReviewSettings((reviewSettings) {
+      return reviewSettings.firstApplicationLaunch == null
+          ? reviewSettings.copyWith(
```
Comment (jeanplevesque) [inline lib/src/review_service/review_service.extensions.dart:11]: Does this work with custom settings types? I think that it might not because `copyWith` will not return the correct (derived) type.

**Review by jeanplevesque — CHANGES_REQUESTED** [review]

File: lib/src/review_service/review_service.extensions.dart (hunk 18 above)
Comment (jeanplevesque) [reply lib/src/review_service/review_service.extensions.dart:11]: In .NET, the `reviewSettings` variable is a generic type that inherits a specific **record**. Therefore, the `with` operator will return the correct derived type. This isn't equivalent to the dart code here.

Comment (Lee31416) [reply lib/src/review_service/review_service.extensions.dart:11]: Do you mean the application launched or the update review settings method ?

Comment (Lee31416) [reply lib/src/review_service/review_service.extensions.dart:11]: If you meant the application launched tracking I retested everything and it all works

Comment (jeanplevesque) [reply lib/src/review_service/review_service.extensions.dart:11]: You need to test it with a derived class. You need to test with `MyCustomReviewSettings` instead of `ReviewSettings`.

Comment (Lee31416) [reply lib/src/review_service/review_service.extensions.dart:11]: I tested it and it works, althought you gotta specify the customReviewSettings everywhere

Comment (jeanplevesque) [reply lib/src/review_service/review_service.extensions.dart:11]: Hmm what do you mean by that? Can you write an automated test that ensures that custom property values aren't lost when using a custom `ReviewSettings` type and the `trackApplicationLaunched` method?

Comment (Lee31416) [reply lib/src/review_service/review_service.extensions.dart:11]: I added documentation in the readme on how to do it properly after our meeting

File: README.md [hunk 19]
```diff
@@ -1,39 +1,362 @@
...
+Here is a simple code that does dependency injection using `get_it`.
+
+```dart
+   GetIt getIt = GetIt.instance;
+
+   // Register the ReviewPrompter implementation with GetIt
+   getIt.registerSingleton<ReviewPrompter>(ReviewPrompter());
+```
+
+### Adding your CustomReviewSettings
```
Comment (jeanplevesque) [inline README.md]: Explain why one would do this.

File: README.md [hunk 20]
```diff
@@ -1,39 +1,362 @@
...
+        requestCount: super.requestCount,
+        lastRequest: super.lastRequest,
+        favoriteJokesCount: favoriteCount,
+      );
+    }
+  }
+```
+
+Notes : 
+  1. It's important you define a copyWith that overrides the one from the ReviewSettings class while returning the newly added property. This ensures that the calls made to the copyWith method of ReviewSettings will be overriden by your implementation
```
Comment (jeanplevesque) [inline README.md]: use `code` formating when you reference code elements (such as `copyWith`, `ReviewSettings`, etc.)

File: README.md [hunk 21]
```diff
@@ -1,39 +1,362 @@
...
+  2. You need to define a copyWith method that returns the parent reviewSettings value along with your new property passed as a parameter. This method will be used when you will track your review settings further down the line.
```
Comment (jeanplevesque) [inline README.md]: I'd rephrase this because `returns the parent reviewSettings value` could be misinterpreted into returning the parent instance (and not returning a copy).

File: README.md [hunk 22]
```diff
@@ -1,39 +1,362 @@
...
+      );
+    }
+  }
+```
+
+Notes : 
+  1. It's important you define a copyWith that overrides the one from the ReviewSettings class while returning the newly added property. This ensures that the calls made to the copyWith method of ReviewSettings will be overriden by your implementation
+  2. You need to define a copyWith method that returns the parent reviewSettings value along with your new property passed as a parameter. This method will be used when you will track your review settings further down the line. 
+
+Once you've defined your new custom review settings you need to add an extension to the ReviewConditionsBuilder like this :
```
Comment (jeanplevesque) [inline README.md]: This is not needed but rather recommended. This makes it so that the API surface is more uniform.

File: README.md [hunk 23]
```diff
@@ -1,39 +1,362 @@
...
+          (reviewSettings, currentDateTime) =>
+              reviewSettings.favoriteJokesCount >= minimumJokesFavorited,
+        ),
+      );
+      return this;
+    }
+  }
+```
+
+Notes:
```
Comment (jeanplevesque) [inline README.md]: For notes, you could use the following format:

> ⚠ Don't forget to do this or that.

Comment (jeanplevesque) [reply README.md]: Example: https://github.com/nventive/Chinook.DynamicMvvm?tab=readme-ov-file#add-child-viewmodels

File: README.md [hunk 24]
```diff
@@ -1,39 +1,362 @@
...
+    ),
+  );
+```
+
+And use it like that :
+```dart
+  await GetIt.I.get<CustomReviewService>().trackFavoriteJokesCount();
+```
+
+now you're all set to track how many jokes are favorited in your app and prompt for a review once your review conditions are meant (3 favorite jokes in this case).
```
Comment (jeanplevesque) [inline README.md]: Capital letter at the start of sentences.

File: README.md (hunk 19 above)
Comment (jeanplevesque) [inline README.md]: I wonder whether we should move this section in the Features section because it might not be a true next steps (because the default `ReviewSettings` type is designed to be usable directly).

**Review by jeanplevesque — CHANGES_REQUESTED** [review]

File: README.md (hunk 19 above)
Comment (Ouillyang) [reply README.md]: I renamed the section to 'Add Custom Tracking Data' which should be clearer

File: README.md (hunk 20 above)
Comment (Ouillyang) [reply README.md]: Done.

File: README.md (hunk 21 above)
Comment (Ouillyang) [reply README.md]: I changed it to: "You need to define a `copyWith` method that returns the superclass `ReviewSettings` value along with your new property passed as a parameter."

Does that work?

File: README.md (hunk 22 above)
Comment (Ouillyang) [reply README.md]: Updated.

File: README.md (hunk 23 above)
Comment (Ouillyang) [reply README.md]: Updated.

File: README.md (hunk 24 above)
Comment (Ouillyang) [reply README.md]: Updated.

File: README.md (hunk 19 above)
Comment (Ouillyang) [reply README.md]: I moved it to features and renamed the section to 'Add Custom Tracking Data' which follows the previous section 'Add Custom Conditions'

File: README.md [hunk 25]
```diff
@@ -1,39 +1,364 @@
...
 
-## Features (to keep as-is, configure or remove)
-- [Mergify](https://mergify.io/) is configured. You can edit or remove [.mergify.yml](/.mergify.yml).
+Before getting started, please read the [Android](https://developer.android.com/guide/playcore/in-app-review) and [iOS](https://developer.apple.com/design/human-interface-guidelines/ratings-and-reviews) application review documentation.
 
-The following is the template for the final README.md file:
+## Getting Started
 
----
+1. Add the `ReviewService` package to your project.
```
Comment (jeanplevesque) [inline README.md]: This should match the name of the package as it is in pubspec.yaml

File: README.md [hunk 26]
```diff
@@ -1,39 +1,364 @@
...
+## Testing
+
+This is what you need to know before testing and debugging this service. Please note that this may change and you should always refer to the [Apple](https://developer.apple.com/documentation/storekit/skstorereviewcontroller/3566727-requestreview#4278434) and [Android](https://developer.android.com/guide/playcore/in-app-review/test) documentation for the most up-to-date information.
+
+### Android TODO : check if this is still true with the in_app_review package
+
+- You can't test this service while debugging the application, the prompt won't show up. To test it, you need to use the [internal application sharing](https://play.google.com/console/about/internalappsharing/) or the internal testing feature in Google Play Console. See [this](https://developer.android.com/guide/playcore/in-app-review/test) for more details.
+- You can't use a Google Suite account on Google Play to review an application because the prompt will not show up.
+
+### iOS TODO : check if this is still true with the in_app_review package
```
Comment (jeanplevesque) [inline README.md]: There seems to be some TODOs here and just below.

File: README.md (hunk 25 above)
Comment (Ouillyang) [reply README.md]: Updated to `review_service`.

File: README.md (hunk 26 above)
Comment (Ouillyang) [reply README.md]: The info seems to be correct, I've removed the TODOs

**Review by jeanplevesque — APPROVED** [review]

**Review by lesterbotello — APPROVED** [review]
